### PR TITLE
add(main): new package dnote

### PR DESCRIPTION
**Pull request by 2096779623**
# how to test
## server
```bash
mkdir -p $PREFIX/var/lib/postgresql
initdb $PREFIX/var/lib/postgresql
pg_ctl -D $PREFIX/var/lib/postgresql start
createdb dnote
DBHost=localhost DBPort=5432 DBName=dnote DBUser=postgres DBPassword="" DBSkipSSL=true WebURL=localhost:3000 dnote-server start
```

## client
```bash
mkdir -p $HOME/.dnote
# your editor here
echo editor: "nano" > ~/.dnote/dnoterc
echo apiEndpoint: http://localhost:3000/api >> ~/.dnote/dnoterc
# Open http://localhost:3000 in your browser then register an account first
dnote login
```

# log
## server
```bash
~ $ DBHost=localhost DBPort=5432 DBName=dnote DBUser=postgres DBPassword="" DBSkipSSL=true WebURL=localhost:3000 dnote-server start

2022/10/29 11:31:21 Performed 0 migrations
2022/10/29 11:31:21 Started background tasks
2022/10/29 11:31:21 Dnote version 2.0.1 is running on port 3000
{"duration":"54ms","level":"info","method":"GET","msg":"incoming request","origin":"","remoteAddr":"192.168.1.99:49506","statusCode":200,"ts":"2022-10-29T11:32:02.428597833Z","ts_unix":1667043122,"uri":"/","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.139 Safari/537.36"}
{"duration":"373ms","level":"info","method":"POST","msg":"incoming request","origin":"","remoteAddr":"127.0.0.1:39152","statusCode":200,"ts":"2022-10-29T11:33:06.031708642Z","ts_unix":1667043186,"uri":"/api/v3/signin","userAgent":"Go-http-client/1.1"}
```
## client
```bash
~ $ dnote login
  Welcome to Dnote Pro (http://localhost:3000)
  [?] email: admin@utermux.dev
  [?] password: 
  ✔ logged in
~ $ dnote add a
  ✔ added to a
  • book name: a
  • created at: Oct 29, 2022 11:33am (UTC)
  • note id: 1
  • note uuid: b8b66d6c-4a6d-473d-8817-84651bf06131

------------------------content------------------------
test

-------------------------------------------------------

```

* I don't know what software to use as the SMTP server, this is optional.

### CODE REVIEW

**Comment by 2096779623** [comment]
> ~There is a tabulation inconsistency in build.sh~

Sorry, because most of the code in build.sh was edited on my computer (using vscode), and part of the code was edited on termux (using vim), thanks for fix it.

**Comment by ravener** [comment]
I'm not sure if you really need all the `export GOARCH` stuff, doesn't termux's build environment already set those to be ready for cross compilation?

But if we need it, why is `arm` not checked for in those conditions?

**Comment by 2096779623** [comment]
> I'm not sure if you really need all the `export GOARCH` stuff, doesn't termux's build environment already set those to be ready for cross compilation?

Where?

> But if we need it, why is `arm` not checked for in those conditions?

The `GOARCH` of the arm architecture is `arm`, no additional settings are required.

**Comment by ravener** [comment]
Well I never see any other Go packages to set the GOARCH variable in their build scripts

**Comment by ravener** [comment]
@2096779623 the packaging is pretty weird though, there's dnote package and the subpackages are pretty useless and they depend on each other.

I would've expected two packages `dnote` and `dnote-server` so you can install the cli and not install the server and vice versa.

**Comment by 2096779623** [comment]
> @2096779623 the packaging is pretty weird though, there's dnote package and the subpackages are pretty useless and they depend on each other.
> 
> I would've expected two packages `dnote` and `dnote-server` so you can install the cli and not install the server and vice versa.

Like [frp](https://github.com/termux/termux-packages/blob/master/packages/frp/), you can change dnote at any time.

**Comment by ravener** [comment]
But that's not how it works right now, if I try to install the client it will install the server too

**Comment by ravener** [comment]
Screenshot_20221030_212541

As you can see `dnote-client` and `dnote-server` are just a few kbs in size, feels like they are empty. And they just depend on the big dnote package anyway, so you still install everything. You cannot install cli without the server.

**Comment by 2096779623** [comment]
> But that's not how it works right now, if I try to install the client it will install the server too

https://github.com/termux/termux-packages/wiki/Creating-new-package#subpackage-dependencies

I'll try adding this to fix that.

A fact: none of the subpackages use `unversioned`.

**Comment by 2096779623** [comment]
> And is it necessary to depend on postgresql? As long as the actual program does not link to postgresql libraries, it should be removed from depends. Database setup should be up to the user to setup, they could even use some cloud database and they wouldn't need to install the whole postgresql server on termux.

https://github.com/termux/termux-packages/commit/aba5b32e25be7b777f511f855056ffdaf9c979dd

File: packages/dnote/dnote-server.subpackage.sh [hunk 1]
```diff
@@ -0,0 +1,2 @@
+TERMUX_SUBPKG_DESCRIPTION="dnote server"
+TERMUX_SUBPKG_INCLUDE="bin/dnote-server, share/doc/${TERMUX_PKG_NAME}/AGPLv3.txt"
```
Comment (Grimler91) [inline packages/dnote/dnote-server.subpackage.sh:2]: This field shouldn't be comma separated, remove the commas and dnote / dnote-server should be put in the subpackages

File: packages/dnote/dnote-client.subpackage.sh [hunk 2]
```diff
@@ -0,0 +1,2 @@
+TERMUX_SUBPKG_DESCRIPTION="dnote server"
```
Comment (Grimler91) [inline packages/dnote/dnote-client.subpackage.sh:1]: Should probably say dnote client here

File: packages/dnote/build.sh [hunk 3]
```diff
@@ -0,0 +1,55 @@
+TERMUX_PKG_HOMEPAGE=https://www.getdnote.com/
+TERMUX_PKG_DESCRIPTION="A simple command line notebook for programmers"
+TERMUX_PKG_LICENSE="AGPL-V3,GPL-3.0"
+TERMUX_PKG_MAINTAINER="2096779623 <admin@utermux.dev>"
+TERMUX_PKG_VERSION=2.0.1
+TERMUX_PKG_SRCURL=https://github.com/dnote/dnote/archive/refs/tags/server-v${TERMUX_PKG_VERSION}.tar.gz
+TERMUX_PKG_SHA256=28cc3bf93b3849b9a4a5e65e531f8a34d6be6048427d924c56ab8c7887676bad
+TERMUX_PKG_BUILD_IN_SRC=true
+TERMUX_PKG_AUTO_UPDATE=true
+TERMUX_PKG_DEPENDS="dnote-server,dnote-client,postgresql"
```
Comment (Grimler91) [inline packages/dnote/build.sh:10]: Probably dependencies here can be removed, dnote-server and dnote-client depend on the mother package automatically, which is what we want

Comment (2096779623) [reply packages/dnote/build.sh:10]: Does it really automatically depend, but why @MrAdityaAlok asked me to add subpackages to depends?

https://github.com/termux/termux-packages/pull/10797#pullrequestreview-983569815

Comment (Grimler91) [reply packages/dnote/build.sh:10]: [subpackages depend on the mother package automatically](https://github.com/termux/termux-packages/blob/master/scripts/build/termux_create_debian_subpackages.sh#L77), unless `TERMUX_SUBPKG_DEPEND_ON_PARENT` is set to false (or really anything other than `unversioned` or `deps`). 

The mother package normally does not depend on the subpackages though. Typical situation is that we have mother package `libfoo` with library, and subpackage `foo` which contains a program linked against `libfoo` (and so the subpackage depends on the mother package).

I would guess (but am not familiar with frp) that it makes more sense to have `TERMUX_PKG_SUGGESTS="frpc, frps"` for frp

Comment (2096779623) [reply packages/dnote/build.sh:10]: https://github.com/termux/termux-packages/actions/runs/3366734259/jobs/5583566310

same result.

**Pull request merged**